### PR TITLE
correct gradient for ARange

### LINKED ISSUE #5790: Return correct gradient in ARange with float dtype (and floatX-typed gradient when dtype is discrete)
Currently, the [implementation](https://github.com/Theano/Theano/blob/372ae95ef9aae402b7f426d159585d26556eb0a1/theano/tensor/basic.py#L5395-L5397) re-uses the dtype of the inptus, which is not correct.

I think there is a helper function to get the right dtype (floatX if dtype in discrete_dtypes, else dtype) somewhere.

### PR DESCRIPTION
**Pull request by Faruk-Ahmed**
fix #5790

### CODE REVIEW

**Comment by nouiz** [comment]
Some travis failure due to flake8:

https://travis-ci.org/Theano/Theano/jobs/280177905#L1265

**Comment by nouiz** [comment]
While investigating the error you told me in your tests, I ended up finishing it. The time consuming part was the investigation, not the code change.

But this change what was said in the original issue, so it will need review by @abergeron or @lamblin

File: theano/tensor/basic.py [hunk 1]
```diff
@@ -5526,17 +5527,25 @@ def connection_pattern(self, node):
...
-        return [start.zeros_like(),
-                DisconnectedType()(),
-                step.zeros_like()]
+
+        if (self.dtype in discrete_dtypes) and (step > 0):
+            return [start.zeros_like(),
+                    DisconnectedType()(),
+                    step.zeros_like()]
+        else:
+            num_steps_taken = outputs[0].shape[0]  # (stop-start)/step
```
Comment (lamblin) [inline theano/tensor/basic.py]: I realize that "number of steps" is ambiguous in that case. In my mind, if the shape was 2 for instance, it meant there was 1 step taken from the starting point.
The comment is not really accurate either: if for instance `step > 0` but `start > stop`, then `outputs[0].shape[0]` will be `1` instead.

File: theano/tensor/basic.py [hunk 2]
```diff
@@ -5526,17 +5527,25 @@ def connection_pattern(self, node):
...
+            return [gz.sum(dtype=config.floatX),
```
Comment (lamblin) [inline theano/tensor/basic.py]: Not `floatX`, if `start.dtype` is float64, for instance, we should keep `gz`'s dtype.

File: theano/tensor/basic.py [hunk 3]
```diff
@@ -5526,17 +5527,25 @@ def connection_pattern(self, node):
...
+                    DisconnectedType()(),
```
Comment (lamblin) [inline theano/tensor/basic.py]: OK for `DisconnectedType` rather than 0.

File: theano/tensor/basic.py [hunk 4]
```diff
@@ -5526,17 +5527,25 @@ def connection_pattern(self, node):
...
+                    (gz * arange(num_steps_taken, dtype=self.dtype)).sum(
+                        dtype=config.floatX)]
```
Comment (lamblin) [inline theano/tensor/basic.py]: Same thing for the dtype here, we should use `floatX` only if `stop.dtype` is discrete.

File: theano/tensor/tests/test_basic.py [hunk 5]
```diff
@@ -5911,6 +5911,20 @@ def test_Op_integers(self):
         assert np.all(f(10, 2, 2) == np.arange(10, 2, 2))
         assert np.all(f(0, 0, 1) == np.arange(0, 0, 1))
 
+    def test_grads(self):
+        def f(start, stop, step):
+            return ARange(start.type.dtype)(start, stop, step)
+
+        rng = np.random.RandomState(utt.fetch_seed())
+        # Due to the random projection, we should not use the exact
+        # point that shage the shape of the output.
```
Comment (lamblin) [inline theano/tensor/tests/test_basic.py]: "change"?

File: theano/tensor/basic.py [hunk 6]
```diff
@@ -5526,17 +5527,25 @@ def connection_pattern(self, node):
 
         return [[True], [False], [True]]
 
-    def grad(self, inputs, grads):
+    def L_op(self, inputs, outputs, grads):
         start, stop, step = inputs
         gz, = grads
         # start and step affect the output values
         # but the outputs are integers so there's
         # no gradient through them
```
Comment (lamblin) [inline theano/tensor/basic.py]: Update the comment, the point of that PR is that it is not always the case.

File: theano/tensor/tests/test_basic.py [hunk 7]
```diff
@@ -5911,6 +5911,20 @@ def test_Op_integers(self):
...
+        for start, stop, step in [(0, 4.9, 1),
+                                  (1, 5.1, 0.5)]:
+            utt.verify_grad(f, [np.float32(start),
+                                np.float32(stop),
+                                np.float32(step)],
```
Comment (lamblin) [inline theano/tensor/tests/test_basic.py]: Use floatX in the test, not float32. This would have caught the issue with the dtype.

**Comment by Faruk-Ahmed** [comment]
continuing on in #6452

**Pull request merged**